File: csharp/ToolGood.Algorithm/Internals/Functions/MathSum/Function_FORECAST.cs
using System;
using System.Collections.Generic;
using ToolGood.Algorithm.Enums;

namespace ToolGood.Algorithm.Internals.Functions.MathSum
{
	internal sealed class Function_FORECAST : Function_N
	{
		public Function_FORECAST(FunctionBase[] funcs) : base(funcs) {
			if (funcs.Length != 3) {
				throw new ArgumentException($"Function '{Name}' requires exactly 3 parameters.");
			}
		}

		public override string Name => "FORECAST";

		public override Operand Evaluate(AlgorithmEngine engine, Func<AlgorithmEngine, string, Operand> tempParameter)
		{
			var xArg = GetNumber(engine, tempParameter, 0);
			if (xArg.IsErrorOrNone) return xArg;
			var x = xArg.NumberValue;

			var yArrayArg = GetArray(engine, tempParameter, 1);
			if (yArrayArg.IsErrorOrNone) return yArrayArg;

			var xArrayArg = GetArray(engine, tempParameter, 2);
			if (xArrayArg.IsErrorOrNone) return xArrayArg;

			var yValues = new List<decimal>();
			foreach (var item in yArrayArg.ArrayValue) {
				if (item.IsNumber) yValues.Add(item.NumberValue);
			}

			var xValues = new List<decimal>();
			foreach (var item in xArrayArg.ArrayValue) {
				if (item.IsNumber) xValues.Add(item.NumberValue);
			}

			if (yValues.Count != xValues.Count || yValues.Count < 2) return FunctionError();

			int n = yValues.Count;

			decimal sumX = 0, sumY = 0;
			for (int i = 0; i < n; i++) {
				sumX += xValues[i];
				sumY += yValues[i];
			}

			var meanX = sumX / n;
			var meanY = sumY / n;

			// 采用中心化两遍法,避免 n*sumX2 - sumX*sumX 的病态消减与溢出
			decimal numerator = 0, denominator = 0;
			for (int i = 0; i < n; i++) {
				var dx = xValues[i] - meanX;
				numerator += dx * (yValues[i] - meanY);
				denominator += dx * dx;
			}

			if (denominator == 0) {
				return Div0Error();
			}
			var slope = numerator / denominator;
			var intercept = meanY - slope * meanX;

			return Operand.Create(intercept + slope * x);
		}
		public override OperandType GetResultType()
		{
			return OperandType.NUMBER;
		}

		internal override void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, string op = null, string val = null)
		{
			funcs[0].GetParameterTypes(noneEngine, result, OperandType.NUMBER);
			funcs[1].GetParameterTypes(noneEngine, result, OperandType.ARRAY);
			funcs[2].GetParameterTypes(noneEngine, result, OperandType.ARRAY);
		}
	}
}

File: java/toolgood.algorithm/src/main/java/toolgood/algorithm/internals/functions/mathSum/Function_FORECAST.java
package toolgood.algorithm.internals.functions.mathSum;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.function.BiFunction;
import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.NoneEngine;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_N;

final class Function_FORECAST extends Function_N {
    public Function_FORECAST(FunctionBase[] funcs) {
        super(funcs);
        if (funcs.length != 3) {
            throw new IllegalArgumentException("Function '" + Name() + "' requires exactly 3 parameters.");
        }
    }

    @Override
    public String Name() {
        return "FORECAST";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        if (funcs.length < 3)
            return ParameterError(1);

        Operand xArg = GetNumber(engine, tempParameter, 0);
        if (xArg.IsErrorOrNone())
            return xArg;
        BigDecimal x = xArg.NumberValue();

        Operand yArrayArg = GetArray(engine, tempParameter, 1);
        if (yArrayArg.IsErrorOrNone())
            return yArrayArg;

        Operand xArrayArg = GetArray(engine, tempParameter, 2);
        if (xArrayArg.IsErrorOrNone())
            return xArrayArg;

        List<BigDecimal> yValues = new ArrayList<>();
        for (Operand item : yArrayArg.ArrayValue()) {
            if (item.IsNumber())
                yValues.add(item.NumberValue());
        }

        List<BigDecimal> xValues = new ArrayList<>();
        for (Operand item : xArrayArg.ArrayValue()) {
            if (item.IsNumber())
                xValues.add(item.NumberValue());
        }

        if (yValues.size() != xValues.size() || yValues.size() < 2)
            return FunctionError();

        double sumX = 0, sumY = 0, sumXY = 0, sumX2 = 0;
        int n = yValues.size();

        for (int i = 0; i < n; i++) {
            double xv = xValues.get(i).doubleValue();
            double yv = yValues.get(i).doubleValue();
            sumX += xv;
            sumY += yv;
            sumXY += xv * yv;
            sumX2 += xv * xv;
        }

        double meanX = sumX / n;
        double meanY = sumY / n;

        double denominator = n * sumX2 - sumX * sumX;
        if (denominator == 0) {
            return Div0Error();
        }
        double slope = (n * sumXY - sumX * sumY) / denominator;
        double intercept = meanY - slope * meanX;

        return Operand.Create(BigDecimal.valueOf(intercept + slope * x.doubleValue()));
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NUMBER;
    }

    @Override
    void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        funcs[0].GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        funcs[1].GetParameterTypes(noneEngine, result, OperandType.ARRAY);
        funcs[2].GetParameterTypes(noneEngine, result, OperandType.ARRAY);
    }
}
